File: kakao.app/js/ked/view/base/knob.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}}

var knob

import kxk from "../../../kxk.js"
let post = kxk.post

import theme from "../../theme/theme.js"

import view from "./view.js"


knob = (function ()
{
    _k_.extend(knob, view)
    function knob (screen, name)
    {
        this["draw"] = this["draw"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        knob.__super__.constructor.call(this,screen,name)
        this.setColor('fg',theme.knob.fg)
        this.setColor('bg',theme.knob.bg)
        this.parentName = this.name.slice(0, -5)
        this.frameSide = 'right'
        this.maxWidth = 68
        this.pointerType = 'ew-resize'
    }

    knob.prototype["onMouse"] = function (event)
    {
        knob.__super__.onMouse.call(this,event)
    
        var col, delta, row

        switch (event.type)
        {
            case 'press':
                if (this.hover)
                {
                    post.emit('pointer','grabbing')
                    this.doDrag = true
                    return {redraw:true}
                }
                break
            case 'drag':
                if (this.doDrag)
                {
                    this.hover = true
                    var _a_ = this.cells.posForEvent(event); col = _a_[0]; row = _a_[1]

                    delta = ((function ()
                    {
                        switch (this.frameSide)
                        {
                            case 'left':
                                return -col

                            case 'right':
                                return col

                        }

                    }).bind(this))()
                    post.emit('pointer','grabbing')
                    if (delta)
                    {
                        post.emit('view.size',this.parentName,this.frameSide,delta)
                    }
                    return {redraw:true}
                }
                this.hover = false
                break
            case 'release':
                if (this.doDrag)
                {
                    if (this.hover)
                    {
                        post.emit('pointer',this.pointerType)
                    }
                    delete this.doDrag
                    return {redraw:true}
                }
                break
        }

        return this.hover
    }

    knob.prototype["draw"] = function ()
    {
        if (!this.hover)
        {
            return
        }
        return this.cells.fill_col(0,0,this.cells.rows - 1,'|',this.color.fg,this.color.bg)
    }

    return knob
})()

export default knob;

File: kakao.app/love/lua/view/base/knob.lua
--[[
    000   000  000   000   0000000   0000000  
    000  000   0000  000  000   000  000   000
    0000000    000 0 000  000   000  0000000  
    000  000   000  0000  000   000  000   000
    000   000  000   000   0000000   0000000  

    handles view resizing
    used by dircol and funcol
--]]


local knob = class("knob", view)
    


function knob:init(name) 
        view.init(self, name)
        
        self:setColor('fg', theme.knob.fg)
        self:setColor('bg', theme.knob.bg)
        
        self.parentName = string.sub(self.name, 1, -6)
        self.frameSide = 'right'
        self.maxWidth = 68
        self.pointerType = 'ew-resize'
        return self
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function knob:onMouse(event) 
        view.onMouse(self, event)
        
        if (event.type == 'press') then 
                if self.hover then 
                    post:emit('pointer', 'grabbing')
                    self.doDrag = true
                    return true
                end
        elseif (event.type == 'drag') then 
                if self.doDrag then 
                    self.hover = true
                    
                    local col, row = unpack(self:eventPos(event))
                    
                    local delta = (function () 
    if (self.frameSide == 'left') then 
    return -col
                            elseif (self.frameSide == 'right') then 
    return col
                            end
end)()
                    
                    post:emit('pointer', 'grabbing')
                    if delta then 
                        post:emit('view.size', self.parentName, self.frameSide, delta)
                    end
                    
                    return true
                end
                
                self.hover = false
        elseif (event.type == 'release') then 
                if self.doDrag then 
                    if self.hover then 
                        post:emit('pointer', self.pointerType)
                    end
                    
                    self.doDrag = nil
                    return true
                end
        end
        
        return self.hover
    end

-- 0000000    00000000    0000000   000   000  
-- 000   000  000   000  000   000  000 0 000  
-- 000   000  0000000    000000000  000000000  
-- 000   000  000   000  000   000  000   000  
-- 0000000    000   000  000   000  00     00  


function knob:draw() 
        if not self.hover then return end
        
        self.cells:fill_col(1, 1, self.cells.rows, '|', self.color.fg, self.color.bg)
        -- log "DRAW" @name
        
        return self.cells:render()
    end

return knob